Part of #21421: replace define with constexpr

janclod opened the pull request:
Part of #21421 

Different refactor (mostly few lines of code per replaced `define`), some unused `#define` deleted, one define (`ADD_CLAMP_BODY`) replaced with a template function.

~I realize that this PR may be too big to review at once, please let me know if I should split it.~
Since this PR is a duplicate of #21705 and it incorporates all the feedback given there, I am leaving as is.

ZehMatt commented:
Please don't close PRs if you start over, instead force push.

janclod commented:
> Please don't close PRs if you start over, instead force push.

That was my initial idea, but I failed :pensive: after a rebase, I did delete the upstream branch, this automatically closed the PR and I could not reopen. What is a good workflow, rebase and then force push? I am not familiar with force-push, can I do this after a rebase?

ZehMatt commented:
> > Please don't close PRs if you start over, instead force push.
> 
> That was my initial idea, but I failed 😔 after a rebase, I did delete the upstream branch, this automatically closed the PR and I could not reopen. What is a good workflow, rebase and then force push? I am not familiar with force-push, can I do this after a rebase?

Force push just means it will overwrite the branch forcefully which means also its rewriting its history. You can always reset a branch to its base by using `git reset --hard upstream/develop` and then force push, then you can start adding commits again.

Broxzier commented:
Push only after you've already made a new commit, because otherwise your branch will be the same as develop (or even behind) and GitHub will automatically close the PR.

janclod commented:
Yeeey :D this time I forced pushed... Thanks for teaching me!

ZehMatt commented on src/openrct2/util/Util.cpp:
```diff
@@ -502,43 +502,40 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
-    else                                                                                                                       \
-    {                                                                                                                          \
-        value += value_to_add;                                                                                                 \
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add, T min_cap, T max_cap)
```
I think it definitely beats the macro use. Also you are correct, we can obtain the limits with numeric_limits<T>, no need to pass them as arguments.

janclod replied:
Happy to see the template approach appreciated :) I will improve based on feedback, until the next push ;)

tupaschoal reviewed (approved):
This LGTM. Should definitely be squash merged due to the confusion on some merges.

Might require a few rebases still due to network version changing (for #21850 and maybe more by the time you see it)

ZehMatt commented on src/openrct2/util/Util.cpp:
```diff
@@ -502,43 +502,42 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
-        value += value_to_add;                                                                                                 \
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add)
+{
+    auto max_cap = std::numeric_limits<T>::max();
```
```suggestion
    auto maxCap = std::numeric_limits<T>::max();
```

ZehMatt commented on src/openrct2/util/Util.cpp:
```diff
@@ -502,43 +502,42 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add)
+{
+    auto max_cap = std::numeric_limits<T>::max();
+    auto min_cap = std::numeric_limits<T>::lowest();
```
```suggestion
    auto minCap = std::numeric_limits<T>::lowest();
```

ZehMatt commented on src/openrct2/util/Util.cpp:
```diff
@@ -502,43 +502,42 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
-    else                                                                                                                       \
-    {                                                                                                                          \
-        value += value_to_add;                                                                                                 \
+template<typename T> constexpr T kAddClampBody(T value, T value_to_add)
```
```suggestion
template<typename T> constexpr T kAddClampBody(T value, T valueToAdd)
```

ZehMatt reviewed (changes requested):
Left a few remarks, we don't do snake_case, the rest seems fine to me.

janclod commented:
@ZehMatt thanks for catching the camel cases, incorporated suggestions.

Also improved further the template function `AddClamp` in `Util.h`, please have a last look, do not expect big surprises.

@tupaschoal also rebased on `develop` to with correct network versioning.

janclod reviewed (commented):
Incorporated changes.

tupaschoal commented:
@Gymnasiast or @ZehMatt this looks ready to go (before another merge conflict)

ZehMatt reviewed (approved):
LGTM, Thanks.
